client.go

```diff
@@ -0,0 +1,449 @@
+	return &Client{clientV1: clientV1, clientV1Alpha: clientV1Alpha}
+}
+
+// Client is a Connect client for the server reflection service.
+type Client struct {
+	clientV1        *reflectClient
+	clientV1Alpha   *reflectClient
+	v1unimplemented atomic.Bool
+}
+
+type reflectClient = connect.Client[reflectionv1.ServerReflectionRequest, reflectionv1.ServerReflectionResponse]
+type reflectStream = connect.BidiStreamForClient[reflectionv1.ServerReflectionRequest, reflectionv1.ServerReflectionResponse]
+
+// ClientStreamOption is an option that can be provided when calling [Client.NewStream].
+type ClientStreamOption interface {
+	apply(*clientStreamOptions)
+}
+
+type clientStreamOptions struct {
+	host    string
+	headers http.Header
+}
+
+// WithRequestHeaders is an option that allows the caller to provide the request headers
+// that will be sent when a stream is created.
+func WithRequestHeaders(headers http.Header) ClientStreamOption {
+	return &withRequestHeaders{headers: headers}
+}
+
+type withRequestHeaders struct {
+	headers http.Header
+}
+
+func (w *withRequestHeaders) apply(options *clientStreamOptions) {
+	options.headers = w.headers
+}
+
+// WithReflectionHost is an option that allows the caller to provide the hostname that
+// will be included with all requests on the stream. This may be used by the server
+// when deciding what source of reflection information to use (which could include
+// forwarding the request message to a different host).
+func WithReflectionHost(host string) ClientStreamOption {
+	return &withReflectionHost{host: host}
+}
+
+type withReflectionHost struct {
+	host string
+}
+
+func (w *withReflectionHost) apply(options *clientStreamOptions) {
+	options.host = w.host
+}
+
+// NewStream creates a new stream that is used to download reflection information from
+// the server. This is a bidirectional stream, so can only be successfully used over
+// HTTP/2. The [ClientStream.Close] method should be called when the caller is done
+// with the stream.
+//
+// If any operation returns an error for which [IsReflectionStreamError] returns true,
+// the stream is broken and all subsequent operations will fail. If the error is not
+// a permanent error, the caller should create another stream and try again.
+func (c *Client) NewStream(ctx context.Context, options ...ClientStreamOption) *ClientStream {
+	clientStream := &ClientStream{
+		ctx:    ctx,
+		client: c,
+	}
+	for _, option := range options {
+		option.apply(&clientStream.clientStreamOptions)
+	}
+	// warm-up the stream
+	clientStream.getStream()
+	return clientStream
+}
+
+// IsReflectionStreamError returns true if the given error was the result of a [ClientStream]
+// failing. If the stream returns an error for which this function returns false, only the
+// one operation failed; the stream is still intact and may be used for subsequent operations.
+func IsReflectionStreamError(err error) bool {
+	var streamErr *streamError
+	return errors.As(err, &streamErr)
+}
```

Comment:
This is typical Go naming, but I wonder if the purpose of the function would be clearer if we used a name that  hints at the purpose a bit more: something like `IsReflectionStreamBroken`, maybe?

Reply:
Sure, that seems reasonable. Done